### DISCUSSION TITLE
Regarding the beamforming effect on the coverage maps

### DISCUSSION THREAD

**Discussion by ekilcioglu** [NVlabs/sionna#331] [Q&A] [answered]
Hello,

I have a question related to the coverage maps of Sionna RT. I am new to the tool and trying to get used to it for the last couple of days. I am currently working on an indoor scenario with 10 m x 10 m dimensions and 3 m height. My aim to use Sionna RT is to observe an environment with 1 transmitter, 1 reconfigurable intelligent surface (RIS) and multiple receivers and see the effect of RIS. However, as I know, Sionna RT does not support RIS yet. That's why, my initial idea was to observe only a transmitter and receivers on the coverage map over the area. My question is as follows:

As far as I understand, the coverage maps are created to show the signal attenuation (path-loss) over the area. However, I want to show how beamforming in the transmitter affects the power levels over this area (Something like Fig.5 below). First, think of it without RIS like the case in Fig.5(a). Can Sionna RT output such a similar figure for me with some adjustments?

Thank you in advance for your answer!

![fig5](https://github.com/NVlabs/sionna/assets/159263210/af59932d-7c2d-4b37-9bd6-9381f44d5d50)

W. -L. Hsu, C. -J. Huang, Y. -C. Chen, S. -C. Lin and S. -F. Chang, "Deployment of Millimeter-Wave Reconfigurable Intelligent Surface in An Indoor Scenario Based on Ray-Tracing Simulation," 2023 XXXVth General Assembly and Scientific Symposium of the International Union of Radio Science (URSI GASS), Sapporo, Japan, 2023, pp. 1-4, doi: 10.23919/URSIGASS57860.2023.10265640.

**Comment by faycalaa** [comment, marked as answer]
Hi,

You can use the `precoding_vec` parameter of [Scene.coverage_map()](https://nvlabs.github.io/sionna/api/rt.html#sionna.rt.Scene.coverage_map). Here is an example:

```py
import sionna
import tensorflow as tf
from sionna.rt import load_scene, PlanarArray, Transmitter

num_cols = 128
scene = load_scene(sionna.rt.scene.etoile)

# Configure antenna array for all transmitters
scene.tx_array = PlanarArray(num_rows=1,
                        num_cols=num_cols,
                        vertical_spacing=0.7,
                        horizontal_spacing=0.5,
                        pattern="tr38901",
                        polarization="V")

# Configure antenna array for all receivers
scene.rx_array = PlanarArray(num_rows=1,
                        num_cols=1,
                        vertical_spacing=0.5,
                        horizontal_spacing=0.5,
                        pattern="dipole",
                        polarization="cross")
# Add a transmitters
tx = Transmitter(name="tx",
            position=[46,-47,30],
            orientation=[0.,0,0])
tx.look_at([-110., 33, 1.5])
scene.add(tx)

# Steering vector
steering_angle = 20. # degrees
steering_angle = steering_angle*np.pi/180. # radian
precoding_vec = tf.range(num_cols, dtype=tf.float32)
precoding_vec = tf.exp(tf.complex(0., 2.*np.pi*precoding_vec*0.5*tf.math.sin(steering_angle)))
precoding_vec = precoding_vec / tf.linalg.norm(precoding_vec)

# Compute coverage map
# Set max_depth = 0 for a better visualization
cm = scene.coverage_map(cm_cell_size=[1.,1.], num_samples=int(1e7), precoding_vec=precoding_vec, max_depth=0)

scene.cameras['scene-cam-0'].position = [-500, 213, 400]
scene.cameras['scene-cam-0'].look_at([-45, 7, 1.5])
scene.render('scene-cam-0', coverage_map=cm);
```

![precoding_cm](https://github.com/NVlabs/sionna/assets/43564757/38214a04-017e-4e5e-a2fb-76b6d0523d0e)


Hope this helps.

**Reply by fightingsoul042** [reply]
@faycalaa Could you also share how this would work with MxN planner array to perform 3D beam steering, i.e., steering in elevation and azimuth angle both?

**Comment by ekilcioglu** [comment]
Thank you for your detailed answer! I will consider doing it and let you know if I face a problem. Thanks again!

**Comment by PhilSoy** [comment]
Related to this question... so far I have been using an single element transmitter with an omni pattern without asking myself too many questions. I would like now to use antenna arrays with several antennas elements. I understand from the example that the precoding vector can be included in 'coverage_map', but how should I implement pre-coding in case I just want to emit a signal with the antenna array?
What happens if I define a 'scene.tx_array' with several antenna elements and a 'transmitter' activated with 'scene.add' without specifying any pre-coding vector ?

**Reply by faycalaa** [reply]
Hi,

- When computing coverage maps, not setting a precoding vector results in using a default one equal to `(1/num_tx_ant)*[1, ..., 1]`, [see the related documentation](https://nvlabs.github.io/sionna/api/rt.html#sionna.rt.Scene.coverage_map).

- When computing paths (using [Scene.compute_paths](https://nvlabs.github.io/sionna/api/rt.html#sionna.rt.Scene.compute_paths)), setting up multiple transmit and/or receive antenna elements will result in computing a path coefficient for every pair of transmit-receive antenna. The `num_rx_ant` and `num_tx_ant` dimensions of [`Paths.a`](https://nvlabs.github.io/sionna/api/rt.html#sionna.rt.Paths.a) would then be greater than one. You could then apply a precoding vector (see [this post](https://github.com/NVlabs/sionna/discussions/150#discussioncomment-8994713) for example).

**Reply by PhilSoy** [reply]
Thanks for the clarifications. But there I am still confused. what you report applies to: scene.synthetic_array = False (each pair of antenna is considered separately). in that case it should be several paths coefficient for each antenna pair (not only one as you indicate).
I am currently using the option 'scene.synthetic_array = True'. It seems to me in that case (based on the post) that everything has to be done 'manually' in Python, and the definition of the antenna array (with number of elements, spacing, etc...) is ignored in the rest of the script. In summary, it seems to me that I can define my TX and RX with a single (omni) antenna element and do the processing manually. since Sionna does not provide a specific support. Am I right ?

**Reply by jhoydis** [reply]
Whenever you configure antenna arrays, Sionna RT computes path coefficients between all pairs of antennas. Nothing needs to be done manually.

The property `scene.synthetic_array` only impacts how these coefficients are computed.

For `scene.synthetic_array=False`, ray tracing for all pairs of antennas is done which can be quite memory consuming, especially for large arrays.

 For `scene.synthetic_array=True`, ray tracing is done between the centers of the antenna arrays to compute a single path coefficient for each pair of transmitter and receiver. Then, this path coefficient is synthetically extended to the antenna arrays by applying antenna-specific phase shifts that depended on the angle of departure (transmitter) or on the angle of arrival (receiver). This is explained at the end of the API documentation of [`Scene.coverage_map`](https://nvlabs.github.io/sionna/api/rt.html#sionna.rt.Scene.coverage_map). Note that for the computation of coverage maps, the synthetic array assumption is always made. Here is a screen shot of the relevant part of the documentation:

![image](https://github.com/NVlabs/sionna/assets/5190129/ae04f65a-aca7-4c4a-8e52-09111c1737b1)
